torch/distributed/_tools/memory_tracker.py

```diff
@@ -0,0 +1,246 @@
+        y_2 = [gb for (name, gb) in self.active_memories.values()]
+        y_3 = [gb for (name, gb) in self.reserved_memories.values()]
+        min_val = min(y_1 + y_2 + y_3)
+        max_val = max(y_1 + y_2 + y_3)
+        x = list(i for i in range(len(y_1)))
+        fig = plt.figure(figsize=(16, 8))
+        plt.plot(x, list(y_1), label="allocated_memory")
+        plt.plot(x, list(y_2), label="active_memory")
+        plt.plot(x, list(y_3), label="reserved_memory")
+        plt.xlabel("# Operator Calls")
+        plt.ylabel("Memory (MB)")
+        for marker_name, marker in self.markers.items():
+            if marker_name == "fw_bw_boundary":
+                plt.plot(
+                    [marker, marker], [min_val, max_val], "r", lw=2, label=marker_name
+                )
+            else:
+                plt.plot(
+                    [marker, marker], [min_val, max_val], "k-", lw=2, label=marker_name
+                )
+        plt.legend()
+
+    def _create_pre_forward_hook(self, name: str) -> Callable:
+        """
+        The pre_foward_hook is to insert current module name with forward prefix for the operator
+        name, also it inserts the marker "fw_start" when the forward pass begins.
+        """
+
+        def _pre_forward_hook(module: nn.Module, inputs: Any) -> None:
+            self.cur_module = f"{name}.forward"
+            if hasattr(module, "is_root") and module.is_root:
+                self._add_marker("fw_start")
+
+        return _pre_forward_hook
+
+    def _create_post_forward_hook(self, name: str) -> Callable:
+        """
+        The post_forward_hook inserts the marker "fw_bw_boundary" at the boundary
+        of forward pass and backward pass.
+        """
+
+        def _post_forward_hook(
+            module: nn.Module,
+            inputs: Sequence[torch.Tensor],
+            outputs: Sequence[torch.Tensor],
+        ) -> None:
+            if hasattr(module, "is_root") and module.is_root:
+                self._add_marker("fw_bw_boundary")
+
+        return _post_forward_hook
+
+    def _create_backward_hook(self, name: str) -> Callable:
+        """
+        The backward_hook inserts the current module name with backward prefix for the operator name.
+        """
+
+        def _backward_hook(
+            module: nn.Module, grad_input: torch.Tensor, grad_output: torch.Tensor
+        ) -> None:
+            self.cur_module = f"{name}.backward"
+
+        return _backward_hook
+
+    @no_type_check
+    def _record_memory_stats(self, fn_name: str) -> None:
+        """
+        Record current allocated memory, current active memory and current reserved memory,
+        the memory stats dict is indexed with 'self.op_id'.
+        """
+        allocated_memory: float = torch.cuda.memory_allocated() / MB
+        reserved_memory: float = torch.cuda.memory_reserved() / MB
+        active_memory: float = (
+            torch.cuda.memory_stats().get("active_bytes.all.current", 0) / MB
+        )
```

Comment:
Besides these global stats, it will also be helpful to tell how much memory is activation, how much is temporary, etc.

Reply:
Those will be added in followup PRs